### DISCUSSION TITLE
How do I show Debug-Adapter messages to theia user?

### DISCUSSION THREAD

**Discussion by tsiyona** [eclipse-theia/theia#7786] [General] [answered]
Hi,

I am developing a debug-adapter that communicates with theia over a tcp socket.
How can I display to the user  (i.e. showMessage()) certain error messages from the debug adapter?

**Background:**
I spawn a single node.js process which runs my debug-adapter server that listens on a tcp port.
On each connect request from @theia/debug on this port, my tcp server constructs a new MyDebugSession and lets it communicate DAP messages to/from @theia/debug DebugAdapterSession.

I would like to be able to display error messages to theia user from MyDebugSession and from the debug adapter server, if possible.

I tried several ways:
1) I send a DAP error response to theia.
	**Result**: If I open 'Debug adapters' in the Output view and enable DAP messages trace, I can see my error message.
	See yellow rectangle in attached figure-1.

![figure-1](https://user-images.githubusercontent.com/9335329/81499721-4bc3c800-92d6-11ea-8f74-5a7c177e3313.png)

2) I send a DAP 'output' event with the error I want to display.
	**Result**: 
	- The even with the error message is displayed in the Output view if DAP trace is enabled. See orange rectangle in attached figure-1.
	- The error message is displayed in  theia Debug Console, if it is open. See attached figure-2.

![figure-2](https://user-images.githubusercontent.com/9335329/81499732-5bdba780-92d6-11ea-8b9e-6f9c4cfd3df8.png)

This is not how I want to see the message.
I would like my message to be displayed in a message pop-up window like this one:

![figure-3](https://user-images.githubusercontent.com/9335329/81500001-e83a9a00-92d7-11ea-872e-d1d1d5a175c0.png)

I saw that only failure to process 'attach' DAP request results in showMessage() of the error message.

Thank you.

**Comment by akosyakov** [comment]
Is it something which should be aligned with VS Code? If it is so, then please file a bug request.

If not, I don't think it is easily doable right now with Theia extensions. Generally we recommend to use VS Code extensions to provide language and debug support. VS Code extension API allows to install message tracker: https://github.com/microsoft/vscode/blob/44bcbe587d7f265ab0457c7931a7baab2419ad66/src/vs/vscode.d.ts#L10459 Maybe you can use it to intercept messages and add additional logic.

**Comment by tsiyona** [comment, marked as answer]
No, it is not something that should be aligned with VS Code.

We wrote a theia extension which contributes a 'DebugAdapterContribution' that does rely on VS Code code, as follows:

![debug-adapter-contribution](https://user-images.githubusercontent.com/9335329/81810627-f1bc4000-952b-11ea-937e-67060484ae05.png)

Then we start a server listening on a tcp port (3456, in the example), and thus serves theia debug requests.

When we start that debug server, I would like to notify theia user upon errors (eg. failure to connect to our backend). 
I would like to notify theia user about fatal errors ocurring during a debug session as well.
The only means I have now to do so are error DAP responses or DAP 'output' event. In both cases, the error message is not displayed to the user.

I saw that DAP DebugProtocol.Message has an attribute called 'showUser'. What's the purpose of this attribute?

**Reply by vince-fugnitto** [reply]
> I saw that DAP DebugProtocol.Message has an attribute called 'showUser'. What's the purpose of this attribute?

I see one example of [`showUser`](https://github.com/microsoft/vscode/blob/77b5058284a50445674ee3b237ad64793b38965e/src/vs/workbench/contrib/debug/browser/rawDebugSession.ts#L672-L674) in vscode itself which uses this attribute to display error notifications to end-users. I do not see a similar use of `showUser` in Theia however.

```ts
private handleErrorResponse(errorResponse: DebugProtocol.Response): Error { 
...
  if (error && error.format && error.showUser) {
    this.notificationService.error(userMessage);
  }
...
}

**Reply by akosyakov** [reply]
@vince-fugnitto Could you file an issue please? Does not sound hard to fix.

**Reply by vince-fugnitto** [reply]
> @vince-fugnitto Could you file an issue please? Does not sound hard to fix.

I've opened the following issue: https://github.com/eclipse-theia/theia/issues/7808

**Reply by tsiyona** [reply]
Thank you @vince-fugnitto  and @akosyakov .
Looking forward to having this feature added to Theia.

**Reply by akosyakov** [reply]
PRs are welcomed :)
